File: Python/src/MergeSort.py
def mergeSort(A, p, r):
    if p >= r:
        return
    q = (p + r) // 2
    mergeSort(A, p, q)
    mergeSort(A, q+1, r)
    merge(A, p, q, r)

def merge(A, p, q, r):
    nL = q - p + 1
    nR = r - q
    L = A[p:q+1]
    R = A[q+1:r+1]

    i = 0
    j = 0
    k = p

    while i < nL and j < nR:
        if L[i] < R[j]:
            A[k] = L[i]
            i += 1
        else:
            A[k] = R[j]
            j += 1
        k += 1
    
    while i < nL:
        A[k] = L[i]
        i += 1; 
        k += 1

    while j < nR:
        A[k] = R[j]
        j += 1; 
        k += 1

File: CSharp/src/Sorts/MergeSort.cs
namespace Algorithm.Sorts;

public class MergeSort
{
    public void Merge(int[] array, int left, int middle, int right)
    {
        var nL = middle - left + 1; // length of Array[left ... middle]
        var nR = right - middle; // length of Array[middle + 1 ... right]
        
        var leftArray = array[left..(middle+1)];
        var rightArray = array[(middle+1)..(right+1)];
        
        var i = 0;
        var j = 0;
        var k = left;

        while (i < nL && j < nR)
        {
            if (leftArray[i] < rightArray[j])
            {
                array[k] = leftArray[i];
                i++;
            }
            else
            {
                array[k] = rightArray[j];
                j++;
            }

            k++;
        }

        while (i < nL)
        {
            array[k] = leftArray[i];
            k++;
            i++;
        }
  
        while (j < nR)
        {
            array[k] = rightArray[j];
            k++;
            j++;
        }
    }

    public void Sort(int[] array, int left, int right)
    {
        if (left >= right) 
            return;
        var middle = (left + right) / 2;
        Sort(array, left, middle);
        Sort(array, middle + 1, right);
        Merge(array, left, middle, right);
    }
}